Firestore/core/src/firebase/firestore/api/collection_reference.mm

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/api/collection_reference.h"
+
+#include <utility>
+
+#import "Firestore/Source/Core/FSTQuery.h"
+
+#include "Firestore/core/src/firebase/firestore/api/document_reference.h"
+#include "Firestore/core/src/firebase/firestore/api/input_validation.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+#include "Firestore/core/src/firebase/firestore/util/autoid.h"
+#include "Firestore/core/src/firebase/firestore/util/hashing.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+
+namespace util = firebase::firestore::util;
+using firebase::firestore::api::ThrowInvalidArgument;
+using firebase::firestore::model::DocumentKey;
+using firebase::firestore::model::ResourcePath;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+namespace {
+
+FSTQuery* MakeQuery(model::ResourcePath path) {
+  if (path.size() % 2 != 1) {
+    ThrowInvalidArgument("Invalid collection reference. Collection references "
+                         "must have an odd number of segments, but %s has %s",
+                         path.CanonicalString(), path.size());
+  }
+
+  return [FSTQuery queryWithPath:path];
+}
+
+}  // namespace
+
+CollectionReference::CollectionReference(model::ResourcePath path,
+                                         std::shared_ptr<Firestore> firestore)
+    : Query(MakeQuery(path), std::move(firestore)) {
+}
+
+bool operator==(const CollectionReference& lhs,
+                const CollectionReference& rhs) {
+  return lhs.firestore() == rhs.firestore() &&
+         objc::Equals(lhs.query(), rhs.query());
+}
+
+size_t CollectionReference::Hash() const {
+  return util::Hash(firestore().get(), query());
```

Comment:
Essentially it comes down to the difference behind API classes vs implementation classes.

API classes are meant to be useful to end users. They're designed to be "live": you can navigate e.g. from a `FIRQuery` to the `FIRFirestore` that it will run against.

Implementation classes are meant to be immutable and context free.

The difference affects how we test: implementation classes are easy to unit test and understand their dependencies. API classes, by virtue of ultimately needing a Firestore instance to back them, are tricky to instantiate and test, except by just spinning up the whole system.